ggml-metal.metal

```diff
@@ -1921,7 +1411,7 @@ kernel void kernel_get_rows(
 #define SG_MAT_ROW 8
 
 // each block_q contains 16*nl weights
-template<typename block_q, short nl, void (*dequantize_func)(device const block_q *, short, thread half4x4 &)>
+template<typename block_q_type, short nl, template<typename, typename, typename> class quant_dri>
 kernel void kernel_mul_mm(device const  uchar * src0,
                            device const  float * src1,
                            device        float * dst,
```

Comment:
The body of this template has to be simplified by factoring out the code that does not depend on the quantization type into separate function calls. We should try to avoid the template by having separate inline implementations of the kernels for each quantization, reusing the common functions and calling the respective `dequantize_` functions.